dimod/cyvariables.pyx

```diff
@@ -0,0 +1,263 @@
+
+        Raises:
+            ValueError: If a variable is present and `permissive` is
+            False.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        # todo: performance improvements for range etc
+        for v in iterable:
+            self._append(v, permissive=permissive)
+
+    def _pop(self):
+        """Remove the last variable.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        if not self:
+            raise IndexError("Cannot pop when Variables is empty")
+
+        self._stop = idx = self._stop - 1
+
+        label = self._index_to_label.pop(idx, idx)
+        self._label_to_index.pop(label, None)
+        return label
+
+    def _relabel(self, mapping):
+        """Relabel the variables in-place.
+
+        Args:
+            mapping (dict):
+                Mapping from current variable labels to new, as a dict. If
+                an incomplete mapping is specified, unmapped variables keep
+                their current labels.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        for submap in iter_safe_relabels(mapping, self):
+            for old, new in submap.items():
+                if old == new:
+                    continue
+
+                idx = self._label_to_index.pop(old, old)
+
+                if new != idx:
+                    self._label_to_index[new] = idx
+                    self._index_to_label[idx] = new  # overwrites old idx
+                else:
+                    self._index_to_label.pop(idx, None)
+
+    def _relabel_as_integers(self):
+        """Relabel the variables as integers in-place.
+
+        Returns:
+            dict: The mapping that would restore the original labels.
+
+        This method is semi-public. it is intended to be used by
+        classes that have :class:`.Variables` as an attribute, not by the
+        the user.
+        """
+        mapping = self._index_to_label.copy()
+        self._index_to_label.clear()
+        self._label_to_index.clear()
+        return mapping
+
+    cdef Py_ssize_t _count_int(self, object v) except -1:
+        # only works when v is an int
+        cdef Py_ssize_t vi = v
+
+        if self._is_range():
+            return 0 <= vi < self._stop
+
+        # need to make sure that we're not using the integer elsewhere
+        return (0 <= vi < self._stop
+                and not PyDict_Contains(self._index_to_label, v)
+                or PyDict_Contains(self._label_to_index, v))
+
+    cpdef Py_ssize_t count(self, object v) except -1:
+        """Return the number of times `v` appears in Variables.
```

Comment:
Either "in the variables" or "in the :class:`.Variables` class", otherwise it looks 17th-century